From ad2f6462b18f5f8e3f2a8a6e28dc549208a3ebfc Mon Sep 17 00:00:00 2001
From: Zhen Lei <thunder.leizhen@huawei.com>
Date: Thu, 28 Apr 2022 18:14:00 +0800
Subject: [PATCH 2130/2592] [Backport] of: Enable crashkernel=X,high to support
 kexec option -s

hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I545H8

Reference: https://gitee.com/openeuler/kernel/commit/88d74c39115408d62173aacf5ffec1b8b221341c

--------------------------------

For "crashkernel=X,high", there must be two crash regions: high=crashk_res
and low=crashk_low_res. But now the syscall kexec_file_load() only add the
crashk_res into "linux,usable-memory-range", this causes the second kernel
to have no available dma memory. Fix it like kexec tool do for option -c.

Signed-off-by: Zhen Lei <thunder.leizhen@huawei.com>
Reviewed-by: Kefeng Wang <wangkefeng.wang@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Li Huafei <lihuafei1@huawei.com>
---
 drivers/of/kexec.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/of/kexec.c b/drivers/of/kexec.c
index f335d941a716..d8231c34e873 100644
--- a/drivers/of/kexec.c
+++ b/drivers/of/kexec.c
@@ -396,6 +396,15 @@ void *of_kexec_alloc_and_setup_fdt(const struct kimage *image,
 				crashk_res.end - crashk_res.start + 1);
 		if (ret)
 			goto out;
+
+		if (crashk_low_res.end) {
+			ret = fdt_appendprop_addrrange(fdt, 0, chosen_node,
+					FDT_PROP_MEM_RANGE,
+					crashk_low_res.start,
+					crashk_low_res.end - crashk_low_res.start + 1);
+			if (ret)
+				goto out;
+		}
 	}
 
 	/* add bootargs */
-- 
2.17.1

